crawler/db.py

```diff
@@ -80,24 +89,77 @@ def get_mongo_collection(database: Database, collection_name: str) -> Collection
     return database[collection_name]
 
 
-def copy_collection(database: Database, collection: Collection) -> None:
-    """Copy a collection to a timestamped version of itself.
+def rename_collection_with_suffix(
+    database: Database, collection: Collection, suffix: str = current_time()
+) -> None:
+    """Renames a collection to a timestamped version of itself
 
     Arguments:
-        database {Database} -- the database of the collection to copy
-        collection {Collection} -- the collection to copy
+        database {Database} -- the database of the collection to rename
+        collection {Collection} -- the collection to rename
     """
-    cloned_collection = f"{collection.name}_{datetime.now().strftime('%y%m%d_%H%M')}"
+    new_name = f"{collection.name}_{suffix}"
 
-    logger.debug(f"Copying '{collection.name}' to '{cloned_collection}'")
+    rename_collection(database, collection, new_name)
+    return None
+
+
+def rename_collection(database: Database, collection: Collection, new_name: str) -> None:
+    """Renames a collection to the new name.
+
+    Arguments:
+        database {Database} -- the database of the collection to rename
+        collection {Collection} -- the collection to rename
+        new_name {str} -- the new name of the collection
+    """
+    logger.debug(f"Renaming '{collection.name}' to '{new_name}'")
 
     # get a list of all docs
-    current_docs = list(collection.find())
+    collection.rename(new_name)
+
+    logger.debug(f"Collection renamed to: '{new_name}'")
 
-    result = database[cloned_collection].insert_many(current_docs)
+    return None
 
-    logger.debug(f"{len(result.inserted_ids)} documents copied to '{cloned_collection}'")
 
+@contextmanager
+def safe_collection(
+    database: Database, collection_name: str, timestamp: str
+) -> Iterator[Collection]:
+    """
+    Creates a context which yields a new temporary collection.
+    If the context runs successfully, renames collection_name to collection_name_timestamp
+    and renames the temporary collection to replace collection_name.
+    If the context fails, the original collection is left in place. The temporary collection is not
+    cleaned up to assist with debugging.
+
+    Arguments:
+        database {Database} -- the database of the collection to replace
+        collection {Collection} -- the collection to replace
+        timestamp {str} -- A timestamp to apply to the original and temporary collection names
+    """
+    temporary_collection_name = f"tmp_{collection_name}_{timestamp}"
+    logger.debug(f"Generating temporary collection: {temporary_collection_name}")
+    temporary_collection = get_mongo_collection(database, temporary_collection_name)
```

Comment:
Is get_mongo_collection creating a new collection if it doesnt exist?

Reply:
Sort of, Mongo creates collections lazily as soon as something gets added to them. Its actually a bit frustrating, as it makes it surprisingly tricky to check if a collection already exists or not.